Feat/updating test script

Mohsinsiddi opened the pull request:
Updates:
1.) Updated Test script with reviews on old PR
2.) Updated Makefile
3.) Removed pip based installation and migrated to poetry based installations

OjusWiZard reviewed (changes requested):
Few more comment, almost there.

Mohsinsiddi commented on Makefile:
```diff
@@ -0,0 +1,20 @@
+run_no_staking_tests:
+	poetry run pytest -v tests/test_run_service.py -s --log-cli-level=INFO
+
+# Clean up after tests
+cleanup:
+	rm -rf /Users/siddi_404/Solulab/OLAS/middleware/quickstart/.operate
+	rm -rf logs/
```
We've addressed this by:
- Removed direct .operate folder deletion
- Using tempfile.TemporaryDirectory for test isolation
- Copying project files to temp directory
- Running tests in the temp directory, which safely contains the .operate folder
- Cleaning up temp directory after tests complete

The .operate folder in the user's actual directory is never touched now.

Mohsinsiddi commented on Makefile:
```diff
@@ -0,0 +1,20 @@
+# Run all commands in sequence
+test: install run_no_staking_tests
+
+stop_service:
+	./stop_service.sh
```
Removed the stop_service target from Makefile as it will be handled by the config path in the future. Users will use ./stop_service.sh with config path directly.

Mohsinsiddi commented on pyproject.toml:
```diff
@@ -9,7 +9,28 @@ packages = []
-olas-operate-middleware = { git = "https://github.com/valory-xyz/olas-operate-app.git", rev = "5dcf737147f81ba42ac78efce6b24729eee2cf3b", python = ">=3.9,<3.12" }
+python = ">=3.9,<3.12"
+olas-operate-middleware = { git = "https://github.com/valory-xyz/olas-operate-app.git", rev = "3a26856254e57bff45a7ce1c3daecc7180c63df5", python = ">=3.9,<3.12" }
+pexpect = "^4.9.0"
+web3 = "^6.1.0"
+termcolor = "^2.5.0"
+colorama = "^0.4.6"
+eth-account = "^0.8.0"
+eth-keyfile = "^0.6.1"
+eth-keys = "^0.4.0"
+eth-rlp = "^0.3.0"
+eth-typing = "^3.5.2"
+eth-utils = "^2.3.1"
+hexbytes = "^0.3.1"
+pytest = "^7.2.1"
+aiohttp = "^3.9.5"
+cytoolz = "^0.12.3"
+frozenlist = "^1.4.1"
+multidict = "^6.0.5"
+requests = "^2.31.0"
+packaging = "^23.2"
+python-dotenv = ">=0.14.5,<0.22.0"
+docker = "^6.1.2"
```
Addressed by:
1. Removed all overlapping dependencies that are already included in olas-operate-middleware
2. Added only pexpect as it's the only test-specific dependency not included in olas-operate-middleware and required for our testing
3. Placed pexpect in the dev group so it won't affect normal users

The current pyproject.toml is now minimal and maintainable, with:
- Only the core olas-operate-middleware dependency in the main dependencies
- Only pexpect in the dev dependencies
- No redundant dependencies to maintain

This aligns with the review suggestion of "Only specifying the required ones (which are not in olas-operate-middleware) like maybe pexpect would be sufficient."

Mohsinsiddi commented on tests/test_run_service.py:
```diff
@@ -0,0 +1,376 @@
+        try:
+            cls.logger.info("Starting run_service.py test")
+            
+            # Get current virtual environment path
+            venv_path = os.environ.get('VIRTUAL_ENV')
+            
+            # Temporarily deactivate virtual environment
+            if venv_path:
+                temp_env = os.environ.copy()
+                temp_env.pop('VIRTUAL_ENV', None)
+                temp_env.pop('POETRY_ACTIVE', None)
+                
+                # Update PATH to remove the virtual environment
+                paths = temp_env['PATH'].split(os.pathsep)
+                paths = [p for p in paths if not p.startswith(venv_path)]
+                temp_env['PATH'] = os.pathsep.join(paths)
+            else:
+                temp_env = os.environ
```
We've kept the _setup_environment as a separate classmethod in TestService class because:
- It makes the code more maintainable and reusable
- Other test classes can inherit from TestService and reuse this setup
- It's cleaner to separate environment setup from test-specific setup

Mohsinsiddi commented on tests/test_run_service.py:
```diff
@@ -0,0 +1,376 @@
+        self.logger.info("Testing service health...")
+        assert check_service_health(self.logger) == True, "Health check failed"
+        
+    def test_04_shutdown_logs(self):
+        """Test service shutdown logs"""
+        self.logger.info("Testing shutdown logs...")
+        assert check_shutdown_logs(self.logger) == True, "Shutdown logs check failed"
```
Updated test_04_shutdown_logs to:
- Explicitly stop the service first
- Wait for containers to stop
- Verify containers are gone
- Then check shutdown logs

This ensures proper sequence and verification of shutdown process.

OjusWiZard commented:
it's failing when I ran it locally, is it working on your end? Also, please rebase it on the latest main to resolve the conflicts

Mohsinsiddi commented:
## Changes Made
- Extended wait durations throughout test script to improve test stability
- Added progress spinners with countdown timers to provide better visibility during wait periods
- Removed test cases related to service logs as they are no longer needed

OjusWiZard commented on pyproject.toml:
```diff
@@ -10,9 +10,12 @@ include = []
 [tool.poetry.dependencies]
 python = ">=3.9,<3.12"
-olas-operate-middleware = { git = "https://github.com/valory-xyz/olas-operate-app.git", rev = "caed60d6757f9fa477d9723fd2ec017413ce5d6f"}
+olas-operate-middleware = { git = "https://github.com/valory-xyz/olas-operate-app.git", rev = "caed60d6757f9fa477d9723fd2ec017413ce5d6f", python = ">=3.9,<3.12" }
```
this can be removed I guess
```suggestion
olas-operate-middleware = { git = "https://github.com/valory-xyz/olas-operate-app.git", rev = "caed60d6757f9fa477d9723fd2ec017413ce5d6f"}
```

Mohsinsiddi replied:
removed

OjusWiZard commented on tests/test_run_service.py:
```diff
@@ -0,0 +1,475 @@
+                    logs = container.logs(tail=50).decode('utf-8')
+                    logger.error(f"Container logs:\n{logs}")
+                    
+                    # Try to restart the container if it exited with 0
+                    if exit_code == 0:
+                        logger.info(f"Attempting to restart container {container.name}")
+                        container.start()
```
why do we try to restart the container?

Mohsinsiddi replied:
removed this code block, it was just there to restart incase of any error while starting conatiners in CI

OjusWiZard replied:
any error should fail the test, that's the purpose of the test :)

Mohsinsiddi replied:
noted

OjusWiZard commented on tests/test_run_service.py:
```diff
@@ -0,0 +1,475 @@
+            time.sleep(5 - elapsed)
+    
+    # Return True only if all checks were successful
+    is_healthy = metrics['successful_checks'] == metrics['check_count']
+    return is_healthy, metrics
```
we can directly return True here right? because we're already returning False in all other paths.
I'm concerned because it's possible that it only iterates 23 times because of some lag in between 120 seconds.

Mohsinsiddi replied:
**For health check metrics:
"I've updated the health check function to use a success rate approach instead of exact check count. Now it:**


Tracks actual checks performed
Calculates a success rate
Returns True if success rate ≥ 90%
Logs detailed metrics including success rate
This handles the case where we might get fewer iterations due to lag."

Mohsinsiddi replied:
thats very narrow edge case, bt thats safe hands, lemme update it that way then

Mohsinsiddi replied:
updated

OjusWiZard reviewed (commented):
It works locally, mostly looks good now. I've suggest some cleanups and two few question.

OjusWiZard commented:
also I noticed, it fails in the CI probably because of that `STAKING_CHOICE`. You know how to fix that :)

Mohsinsiddi commented on .github/workflows/python-tests.yml:
```diff
@@ -0,0 +1,136 @@
+        echo "BACKUP_WALLET=${BACKUP_WALLET}" >> .env
+        echo "TEST_PASSWORD=${TEST_PASSWORD}" >> .env
+        echo "PRIVATE_KEY=${PRIVATE_KEY}" >> .env
+        echo "STAKING_CHOICE=${STAKING_CHOICE}" >> .env
```
"Removed all references to STAKING_CHOICE from the workflow file and .env creation as this needs to be handled dynamically in the test script."

Mohsinsiddi commented on .github/workflows/python-tests.yml:
```diff
@@ -0,0 +1,136 @@
+      BACKUP_WALLET: ${{ secrets.BACKUP_WALLET }}
+      TEST_PASSWORD: ${{ secrets.TEST_PASSWORD }}
+      PRIVATE_KEY: ${{ secrets.PRIVATE_KEY }}
+      STAKING_CHOICE: ${{ secrets.STAKING_CHOICE }}
```
"Removed all references to STAKING_CHOICE from the workflow file and .env creation as this needs to be handled dynamically in the test script."

Mohsinsiddi commented on .github/workflows/python-tests.yml:
```diff
@@ -0,0 +1,136 @@
+name: Quickstart Tests
+on:
+  push:
+    branches:
+      - feat/updating_test_script
```
For workflow branch trigger:
"Removed the feat/updating_test_script branch and replaced it with workflow_dispatch for manual triggering when needed."

Mohsinsiddi replied:
please review latest ci yaml file

OjusWiZard commented on Makefile:
```diff
@@ -0,0 +1,12 @@
+
+# Install project dependencies
+install:
+	poetry install
```
`poetry install` is done in every `.sh` file like `run_service.sh`, `stop_service.sh`, etc. So we'll need to update those also like `poetry install --only main` in them, so that the user doesn't bloat their space with dev dependencies.

Mohsinsiddi replied:
sure, updated all *.sh files
